cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/cask360/Cask360TableDataset.java

```diff
@@ -0,0 +1,424 @@
+    byte[][][] arrays = new byte[2][][];
+    int len = data.size();
+    int idx = 0;
+    arrays[0] = new byte[len][];
+    arrays[1] = new byte[len][];
+    for (Map.Entry<byte[], byte[]> entry : data.entrySet()) {
+      arrays[0][idx] = entry.getKey();
+      arrays[1][idx] = entry.getValue();
+      idx++;
+    }
+    return arrays;
+  }
+
+  private static byte[][][] flattenDataArrays(List<byte[][][]> srcArrays) {
+    int count = 0;
+    for (byte[][][] srcArray : srcArrays) {
+      count += srcArray[0].length;
+    }
+    byte[][][] dataArray = new byte[2][][];
+    dataArray[0] = new byte[count][];
+    dataArray[1] = new byte[count][];
+    int dataIdx = 0;
+    for (byte[][][] srcArray : srcArrays) {
+      int length = srcArray[0].length;
+      for (int i = 0; i < length; i++) {
+        dataArray[0][dataIdx] = srcArray[0][i];
+        dataArray[1][dataIdx] = srcArray[1][i];
+        dataIdx++;
+      }
+    }
+    return dataArray;
+  }
+
+  // Implementations of RecordScannable
+  // Flattens the table into (id, group, key, value) rather than a complex type
+
+  @Override
+  public Type getRecordType() {
+    return new TypeToken<Cask360Record>() {
+      private static final long serialVersionUID = Cask360Record.serialVersionUID;
+    }.getType();
+  }
+
+  @Override
+  public List<Split> getSplits() {
+    return dataTable.getSplits();
+  }
+
+  @Override
+  public RecordScanner<Cask360Record> createSplitRecordScanner(Split split) {
+    return new RecordScanner<Cask360Record>() {
+      private SplitReader<byte[], Row> splitReader;
+      private String id = null;
+      private Iterator<KeyValue<byte[], byte[]>> columnIterator = null;
+      private Iterator<Cask360Record> recordIterator = null;
+
+      @Override
+      public void initialize(Split split) throws InterruptedException {
+        this.splitReader = dataTable.createSplitReader(split);
+        this.splitReader.initialize(split);
+      }
+
+      @Override
+      public boolean nextRecord() throws InterruptedException {
+        // Check if the value iterator is active
+        if (recordIterator != null) {
+          if (recordIterator.hasNext()) {
+            return true;
+          } else {
+            recordIterator = null;
+          }
+        }
+        if (id == null) {
+          // Initial call
+          if (!this.splitReader.nextKeyValue()) {
+            return false;
+          }
+          id = Bytes.toString(this.splitReader.getCurrentKey());
+          columnIterator = columnIterator(this.splitReader.getCurrentValue());
+          return columnIterator.hasNext();
+        } else if (!columnIterator.hasNext()) {
+          // When current iterator is done
+          if (!this.splitReader.nextKeyValue()) {
+            return false;
+          }
+          id = Bytes.toString(this.splitReader.getCurrentKey());
+          columnIterator = columnIterator(this.splitReader.getCurrentValue());
+          return columnIterator.hasNext();
```

Comment:
This `if-else` is pretty duplicate. Also I think it may falsely thinks that the end of the iteration if it encounter an empty columnIterator, even there may be more rows. I think the logic should be something like this.

```java
while (columnIterator == null || !columnIterator.hasNext()) {
  if (!splitReader.nextKeyValue()) {
    break;
  }
  id = Bytes.toString(splitReader.getCurrentKey());
  columnIterator = createColumnIterator(splitReader.getCurrentValue());
}
return columnIterator != null && columnIterator.hasNext();
```